src/current/v23.1/connection-pooling.md

```diff
@@ -25,37 +25,149 @@ Each time an application reads or writes data, it will request one of the connec
 
 Connection pooling can be a enabled as a feature of the driver, a separate library used in conjunction with a driver, a feature of an application server, or a proxy server that acts as a gateway to the database server.
 
+Creating the appropriate size pool of connections is critical to gaining maximum performance in an application. Too few connections in the pool will result in high latency as each operation waits for a connection to open up. But adding too many connections to the pool can also result in high latency as each connection thread is being run in parallel by the system. The time it takes for many threads to complete in parallel is typically higher than the time it takes a smaller number of threads to run sequentially.
+
 {{site.data.alerts.callout_success}}
 To read more about connection pooling, see our [What is Connection Pooling, and Why Should You Care](https://www.cockroachlabs.com/blog/what-is-connection-pooling/) blog post.
 {{site.data.alerts.end}}
 
-## Sizing connection pools
+## Recommended settings for connection pools
 
-Idle connections in CockroachDB do not consume many resources compared to PostgreSQL. Cockroach Labs estimates the memory overhead of idle connections in CockroachDB is 20 kB to 30 kB per connection.
+These are Cockroach Labs recommendations for settings common to most connection pooling software. If your application has multiple workers, each worker should be configured based on its own connections and workload.
 
-Creating the appropriate size pool of connections is critical to gaining maximum performance in an application. Too few connections in the pool will result in high latency as each operation waits for a connection to open up. But adding too many connections to the pool can also result in high latency as each connection thread is being run in parallel by the system. The time it takes for many threads to complete in parallel is typically higher than the time it takes a smaller number of threads to run sequentially.
+### Set the maximum lifetime of connections
+
+The maximum lifetime of a connection should be set to no less than 5 minutes and no more than 30 minutes. Applications that connect to {{ site.data.products.serverless }} clusters should set this to 30 minutes, while applications that connect to {{ site.data.products.dedicated }} and {{ site.data.products.core }} clusters should set this to 5 minutes.
+
+### Set the maximum number of open connections
+
+The maximum number of open connections should be set to a value that is greater than the maximum number of concurrent queries that your application worker expects to make. Doing so minimizes the need to establish new connections.
```

Comment:
> The maximum number of open connections should be set to a value that is greater than the maximum number of concurrent queries that your application worker expects to make. Doing so minimizes the need to establish new connections.

Wait, that's not right: the connection pool acts as a governor, too, to prevent the target cluster from being overwhelmed with active queries.  How about:

> The maximum number of open connections should be scaled based on the workload and capabilities of the target cluster.  The sweet spot will be the minimum of a) the number of vCPUs and IOPS the cluster can perform, times some scaling factor (suggest starting with 4), and b) the maximum number of concurrent queries all workloads targeting this cluster expect to make.  Doing so minimizes the need to establish new connections and keeps the target cluster safe.

If we want to throw in a formula, how about something like this:

* `scaling_factor = 4`
* `num_vcpus_per_node = 16`
* `num_nodes_in_crdb_cluster = 6`
* `num_active_connections = ${num_vcpus_per_node} * ${num_nodes_in_crdb_cluster} * ${scaling_factor} # 384`

If we plug in the avg query time, we can figure out the avg qps:

* `avg_query_time = 10ms`
* `est_num_qps = ${num_active_connections} / ${avg_query_time} # est 38400`

From there, you can add headroom by scaling up the cluster for burst capacity.

* `est_num_qps = 38400`
* `est_num_requests_per_worker = 1600`
* `worker_target_utilization = 0.8`
* `est_num_client_workers = ${est_num_qps} / ${est_num_requests_per_worker} / ${worker_target_utilization} # est 30 workers`

And a max connections per worker of:

* `conn_burst_capacity = 1.5`
* `max_conns = (${num_active_connections} / ${est_num_client_workers}) * ${conn_burst_capacity}`

This gives us a recommendation of about 20 max connections per worker to prevent overload of the target cluster and ensure available connections on every worker to handle this workload.

If some enterprising individual wanted to turn this into an interactive JavaScript calculator presented as a table with various default values, ❤️ very much.

Reply:
None of that works for Serverless, though, since there are not fixed number of vCPUs and nodes. I think this section should have solid, basic advice that's no more than 1-2 sentences. Further down the page we should have more detailed advice that explains "advanced" configuration for optimizing app throughput. I was envisioning two sections with names like: "Configure the Connection Pool" and "Optimize the Connection Pool".

The existing docs page is just overwhelming, with so much detail that I fear customer's eyes will glaze over and they'll end up doing nothing at all.

Reply:
I like @sean-'s suggestion of giving guidance based on the workload characteristics but I agree that it doesn't make any sense to give guidance based on the number of vCPUs for Serverless. I think it is unavoidable that this will morph into having differing guidance based on a per product basis.

I propose that we leave this current section as is which should be good enough for most cases. We can make a separate change for something like "Advanced connection pool sizing guidance" where we go over the details @sean- described. Especially the interactive calculator would be helpful for customers looking to fine tune their setup. I opened an issue to keep track of that here: https://github.com/cockroachdb/docs/issues/17820